pennylane_lightning/lightning_qubit.py

```diff
@@ -131,6 +134,51 @@ def apply_lightning(self, state, operations):
 
         return np.reshape(state_vector, state.shape)
 
+    def adjoint_jacobian(self, tape, starting_state=None, use_device_state=False):
+
+        if self.shots is not None:
+            warn(
+                "Requested adjoint differentiation to be computed with finite shots."
+                " The derivative is always exact when using the adjoint differentiation method.",
+                UserWarning,
+            )
+
+        for m in tape.measurements:
+            if m.return_type is not Expectation:
+                raise QuantumFunctionError(
+                    "Adjoint differentiation method does not support"
+                    f" measurement {m.return_type.value}"
+                )
+
+        # Initialization of state
+        if starting_state is not None:
+            ket = np.ravel(starting_state)
+        else:
+            if not use_device_state:
+                self.reset()
+                self.execute(tape)
+            ket = self._pre_rotated_state
+
+        # TODO: How to accommodate for tensor product observables?
+        adj = AdjointJacobian(ket)
+
+        obs_serialized = _serialize_obs(tape, self.wire_map)
+        ops_serialized = _serialize_ops(tape, self.wire_map)
+
+        jac = adj.adjoint_jacobian(
+            *obs_serialized, *ops_serialized, tape.trainable_params, tape.num_params
```

Comment:
This is still up for debate, but I think we can pass in the tuples directly to CPP, without incurring the expansion overhead.
```suggestion
            obs_serialized, ops_serialized, tape.trainable_params, tape.num_params
```
and I can unpack them in CPP. I have already made some helper struct to simplify data passing around here, so this might just be tidier. I'll let you know once I have it bound up.